src/backend/env/token.rs

```diff
@@ -306,12 +307,22 @@ pub fn transfer(
     Ok(state.ledger.len().saturating_sub(1) as u128)
 }
 
-fn update_user_balance(state: &mut State, principal: Principal, balance: Token) {
+fn update_user_balance(
+    state: &mut State,
+    principal: Principal,
+    balance: Token,
+    minted_tokens: Option<Token>,
+) {
     if let Some(user) = state.principal_to_user_mut(principal) {
         if user.principal == principal {
             user.balance = balance
         } else if user.cold_wallet == Some(principal) {
             user.cold_balance = balance
+        } else {
+            unreachable!("unidentifiable principal")
+        }
+        if let Some(tokens) = minted_tokens {
+            user.minted += tokens;
```

Comment:
This seems inconsistent with the existing balance accounting, which is done per principal.

Would it make sense to track `minted` also per principal? Otherwise, if the user changes the principal, the total minted for the user wouldn't match the total minted for the principal.

Reply:
I was thinking about it more from the design point of  view. Since a balance is just a sum of transactions of a principal, and minted is the sum of mint transactions (so a subset of transactions), it seems cleaner to have the notion of minted balance on the principal as well. 

Tracking it on the user also works, but kind of breaks the symmetry and becomes more ad-hoc IMO.

My opinion is not strong, so this approach is also okay.

Reply:
> it seems cleaner to have the notion of minted balance on the principal as well.

How would we use it though? For example, after this release, new principal changes will not destroy the "minting power" anymore as this field is tracked on the user level. So we're talking right now only about the previously minted minting power, which won't be accounted, but it would not matter as soon as the user reaches the donation cap.

Reply:
Btw, what happens to the 300 cap?!  Does this go to 50 (300 / 6) because of the difficulty factor?!

Reply:
Yes!

Reply:
> How would we use it though? For example, after this release, new principal changes will not destroy the "minting power" anymore as this field is tracked on the user level.

I guess I am trying to say that tracking on the principal level seems more consistent. For example, everyone would be able to verify that their `minted` balance is exactly the sum of balances in their transaction history regardless whether they changed their principal or not. If we track on the user side, then the `minted` balance and their transaction history will diverge after principal changes. So purely from the design point of view, the principal tracking seems more "right".

As I said, we can go ahead with the existing solution since it works as well.

Reply:
I thought more about this and I think I share your point now. It’s indeed more cleaner.  Will change.